smartnews-ads-conversion-api.md

```diff
@@ -275,3 +275,71 @@ Content-Type: application/json
 | Purchase History       | PurchaseHistory                          |
 | Like                   | Like                                     |
 | Install                | Install                                  |
+
+
+## [New] Batch Request
+The Batch Request feature allows advertisers to submit multiple conversions in a single request. All other supported parameters remain the same as described above. Note that only the POST method is supported for batch requests.
+
+### Conversion URL
+Production Endpoint: https://log.smartnews-ads.com/conversion_api/conversions/{api_version}/{partner_name}
+
+Staging Endpoint (test only) https://stg-log.smartnews-ads.com/conversion_api/conversions/{api_version}/{partner_name}
+
+### Example of batch request
+```sh
+curl --request POST 'https://log.smartnews-ads.com/conversion_api/conversions/v1/smartnews' \
+--header 'Content-Type: application/json' \
+--header 'Authorization: {authorization token issued by SN}' \
+--data-raw '{
+  "data": [
+    {
+      "action_source": "app",
+      "event_name": "Purchase",
+      "store_id": "jp.gocro.smartnews.android",
+      "mobile_platform": "Android",
+      "click_id": "UnoPeo4IDmEwnHepAAEA",
+      "event_time": 1473668802,
+      "properties": {
+        "item_id": "akashiro:10003938",
+        "shop_id": "akashiro",
+        "event_value": 100.1,
+        "currency": "JPY",
+        "quantity": 2
+      }
+    },
+    {
+      "action_source": "app",
+      "event_name": "AddToCart",
+      "store_id": "jp.gocro.smartnews.android",
+      "mobile_platform": "Android",
+      "click_id": "UnoPeo4IDmEwnHepAAEA",
+      "event_time": 1473667802,
+      "properties": {
+        "item_id": "akashiro:10003938",
+        "shop_id": "akashiro",
+        "event_value": 100.1,
+        "currency": "JPY",
+        "quantity": 2
+      }
+    },
+    {
+      "action_source": "app",
+      "event_name": "AddToCart",
+      "store_id": "579581125",
+      "mobile_platform": "iOS",
+      "click_id": "UnoFcd4UxewnHepAAEA",
+      "event_time": 1473667802,
+      "properties": {
+        "item_id": "akashiro:10003939",
+        "shop_id": "akashiro",
+        "event_value": 100.1,
+        "currency": "JPY",
+        "quantity": 2
+      }
+    }
+  ]
+}'
+```
+- You can send up to 1,000 conversion events in the data field. However, for optimal performance, it’s recommended to send events as soon as they occur, ideally within one hour of the event.
+- Important: If any invalid events are included in the batch, the entire batch will be rejected.
+- Note: A batch request may respond with a 200 OK status, even if there is an internal error within the batch.
```

Comment:
> A batch request may respond with a 200 OK status, even if there is an internal error within the batch.

Does it mean the client has no way to tell whether the request is successful or not? If so, then it doesn't look like an appropriate design to me.

Besides, I think we need to specify the response format explicitly, rather than `may respond with a 200 OK`. What other responses will be returned in other cases? What about the response body? Such information are missing in the spec.

Reply:
> Does it mean the client has no way to tell whether the request is successful or not? If so, then it doesn't look like an appropriate design to me.

Agree. I actually had a conversation with xuejun-san about the error handling and was told that it was designed this way because he didn’t want the client side to notice internal errors. Instead, the errors would only be visible on our side so we can address them internally.

> Besides, I think we need to specify the response format explicitly, rather than may respond with a 200 OK.

Good point! I think I can reference the example response for this.

> What other responses will be returned in other cases? What about the response body? Such information are missing in the spec.

There are actually some example responses [here](https://github.com/smartnews/smartnews-ads-advertising-api-spec/blob/dc623bf87a7fb4c471cca029be8246a912b2af90/smartnews-ads-conversion-api.md#example-response), but may not be easy to understand, because it's placed under conversion-api not batch-conversion-api.

Reply:
I think batch API deserve it's own response spec because it's not very straight-forward to know what the response will be like compared to single event API. 
E.g. in single event API, we ahve this example.
```
{
    "message": "Bad Request",
    "request_id": "a326f711-1566-4002-9729-2846ae5107c8",
    "error": {
        "issue": "Invalid input provided",
        "detail": "The ‘unit_price’' field data type is incorrect"
    }
}
```
Then, if one of the item in the batch has this issue, will the API return exactly the same response? What about more than 1 item have this issue, or different items have different issues? Will we return only 1 of those issues?

Reply:
Very true! Let me re-create the doc from the private repo and assign it to you later.